server/app/auth/ApiAuthenticator.java

```diff
@@ -0,0 +1,97 @@
+ * database query. This is necessary because pac4j creates profiles in the next step, but we need
+ * the API key to perform authentication (this step).
+ *
+ * <p>Note that at this layer, the request is authenticated, not authorized. All API requests that
+ * reach a controller are already authenticated, but it is the controller's responsibility to check
+ * that the resource being accessed is authorized for the request's API key.
+ *
+ * <p>The API key ID and secret are provided as the basic auth username and password, respectively.
+ * To be valid, a request must reference a valid API key with its username credential, and further
+ * that key must:
+ * <li>Not be retired.
+ * <li>Have an expiration date in the future.
+ * <li>Have a subnet that includes the IP address of the request.
+ * <li>Have a salted key secret that matches the salted password in the request's basic auth
+ *     credentials.
+ */
+public class ApiAuthenticator implements Authenticator {
+
+  private static final int CACHE_EXPIRATION_TIME_SECONDS = 60;
+
+  private final SyncCacheApi syncCacheApi;
+  private final Provider<ApiKeyService> apiKeyService;
+
+  @Inject
+  public ApiAuthenticator(
+      @NamedCache("api-keys") SyncCacheApi syncCacheApi, Provider<ApiKeyService> apiKeyService) {
+    this.syncCacheApi = checkNotNull(syncCacheApi);
+    this.apiKeyService = checkNotNull(apiKeyService);
+  }
+
+  @Override
+  public void validate(Credentials rawCredentials, WebContext context, SessionStore sessionStore) {
+    if (!(rawCredentials instanceof UsernamePasswordCredentials)) {
+      throw new RuntimeException("ApiAuthenticator must receive UsernamePasswordCredentials.");
+    }
+
+    UsernamePasswordCredentials credentials = (UsernamePasswordCredentials) rawCredentials;
+    String keyId = credentials.getUsername();
+
+    // Cache the API key for quick lookup in the controller, also for subsequent requests.
+    // We intentionally cache the empty optional rather than throwing here so that subsequent
+    // requests with the invalid key do not put pressure on the database.
+    ApiKey apiKey =
+        syncCacheApi
+            .getOrElseUpdate(
+                keyId, () -> apiKeyService.get().findByKeyId(keyId), CACHE_EXPIRATION_TIME_SECONDS)
+            .orElseThrow(() -> new BadCredentialsException("API key does not exist"));
+
+    if (apiKey.isRetired()) {
+      throw new BadCredentialsException("API key is retired");
+    }
+
+    if (apiKey.expiredAfter(Instant.now())) {
+      throw new BadCredentialsException("API key is expired");
+    }
+
+    SubnetUtils allowedSubnet = new SubnetUtils(apiKey.getSubnet());
+    // Setting this to true includes the network and broadcast addresses.
+    // I.e. /31 and /32 will not be considered included in the subnetwork
+    // if this is false.
+    allowedSubnet.setInclusiveHostCount(true);
+
+    if (!allowedSubnet.getInfo().isInRange(context.getRemoteAddr())) {
```

Comment:
Probably over engineering at this point, but want to toss something out.

These errors /1